org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java

```diff
@@ -707,7 +708,7 @@ private boolean addConstraintsToC_OneExpr(Expression expri, Set<ConstraintFormul
 			if (innerMethod instanceof ParameterizedGenericMethodBinding)
 				 innerContext = invocation.getInferenceContext((ParameterizedGenericMethodBinding) innerMethod);
 
-			if (innerContext != null) {
+			if (innerContext != null && !innerContext.isInexactVarargsInference()) {
```

Comment:
This is to avoid breaking the test case below.
https://github.com/eclipse-jdt/eclipse.jdt.core/blob/351571327d22c10c6f53943892e0bf4a3c8226ba/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/GenericsRegressionTest_1_8.java#L7133

All other changes in this PR are used to fix the new test cases `testGH506_a` and `testGH506_b`.